sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExecBase.scala

```diff
@@ -198,22 +199,26 @@ class ConditionalNestedLoopJoinIterator(
   }
 
   override def createGatherer(
-      cb: ColumnarBatch,
+      streamBatch: LazySpillableColumnarBatch,
       numJoinRows: Option[Long]): Option[JoinGatherer] = {
     if (numJoinRows.contains(0)) {
       // nothing matched
+      streamBatch.close()
       return None
     }
-    withResource(GpuColumnVector.from(builtBatch.getBatch)) { builtTable =>
-      withResource(GpuColumnVector.from(cb)) { streamTable =>
-        closeOnExcept(LazySpillableColumnarBatch(cb, spillCallback, "stream_data")) { streamBatch =>
-          val builtSpillOnly = LazySpillableColumnarBatch.spillOnly(builtBatch)
-          val (leftTable, leftBatch, rightTable, rightBatch) = buildSide match {
-            case GpuBuildLeft => (builtTable, builtSpillOnly, streamTable, streamBatch)
-            case GpuBuildRight => (streamTable, streamBatch, builtTable, builtSpillOnly)
+    closeOnExcept(streamBatch) { streamBatch =>
+      streamBatch.allowSpilling()
+      withRetryNoSplit {
+        withResource(GpuColumnVector.from(builtBatch.getBatch)) { builtTable =>
+          withResource(GpuColumnVector.from(streamBatch.getBatch)) { streamTable =>
+            val builtSpillOnly = LazySpillableColumnarBatch.spillOnly(builtBatch)
+            val (leftTable, leftBatch, rightTable, rightBatch) = buildSide match {
+              case GpuBuildLeft => (builtTable, builtSpillOnly, streamTable, streamBatch)
+              case GpuBuildRight => (streamTable, streamBatch, builtTable, builtSpillOnly)
+            }
+            val maps = computeGatherMaps(leftTable, rightTable, numJoinRows)
+            makeGatherer(maps, leftBatch, rightBatch, joinType)
```

Comment:
There is a problem here for retries.  makeGatherer can close one of the batches in some cases.  If that happens to the streamBatch, it will not work on a retry.